Converted 1-org readme to new format. (#423)
terraform-google-modules/terraform-example-foundation: 1-org/README.md

<table>
<tbody>
<tr>
<td><a
href="https://github.com/terraform-google-modules/terraform-example-foundation/tree/master/0-bootstrap">0-bootstrap</a></td>
<td>Bootstraps a Google Cloud organization, creating all the required resources
and permissions to start using the Cloud Foundation Toolkit (CFT). This
step also configures a CI/CD pipeline for foundations code in subsequent
stages.</td>
</tr>
<tr>
<td>1-org (this file)</td>
<td>Sets up top level shared folders, monitoring and networking projects, and
organization-level logging, and sets baseline security settings through
organizational policy.</td>
</tr>
<tr>
<td><a
href="https://github.com/terraform-google-modules/terraform-example-foundation/tree/master/2-environments">2-environments</a></td>
<td>Sets up development, non-production, and production environments within the
Google Cloud organization that you've created.</td>
</tr>
<tr>
<td><a
href="https://github.com/terraform-google-modules/terraform-example-foundation/tree/master/3-networks">3-networks</a></td>
<td>Sets up base and restricted shared VPCs with default DNS, NAT (optional),
Private Service networking, VPC service controls, on-premises Dedicated

> Interconnect, and baseline firewall rules for each environment. Also sets

Reviewer: ```suggestion
Interconnect, and baseline firewall rules for each environment. It also sets
```

up the global DNS hub.</td>
</tr>
<tr>
<td><a
href="https://github.com/terraform-google-modules/terraform-example-foundation/tree/master/4-projects">4-projects</a></td>

> <td>Set up a folder structure, projects, and application infrastructure pipeline for applications,

Reviewer: ```suggestion
<td>Sets up a folder structure, projects, and application infrastructure pipeline for applications,
```

 which are connected as service projects to the shared VPC created in the previous stage.</td>
</tr>
</tbody>
</table>

For an overview of the architecture and the parts, see the
[terraform-example-foundation README](https://github.com/terraform-google-modules/terraform-example-foundation).

## Purpose

> The purpose of this step is to [...].

Reviewer 2: ```suggestion
The purpose of this step is to up top level shared folders, monitoring and networking projects, and organization-level logging, and sets baseline security settings through organizational policy.
```

> The purpose of this step is to up top level shared folders, monitoring and networking projects, and organization-level logging, and sets baseline security settings through organizational policy.

Reviewer: ```suggestion
The purpose of this step is to set up top level shared folders, monitoring and networking projects, organization-level logging, and set baseline security settings through organizational policies.
```

## Prerequisites

1. 0-bootstrap executed successfully.
2. Cloud Identity / Google Workspace group for security admins.

> 3. Membership in the security admins group for user running Terraform.

Reviewer: ```suggestion
3. Membership in the security admins group for the user running Terraform.
```

Note: Make sure that you use the same version of Terraform throughout this

> series. Otherwise, you might experience Terraform state snapshot lock errors.

Reviewer: ```suggestion
series, otherwise you might experience Terraform state snapshot lock errors.
```

1. Commit changes.
   ```
   git add .
   git commit -m 'Your message'
   ```
1. Push your plan branch. The branch `plan` is not a special one. Any branch which name is different from `development`, `non-production` or `production` will trigger a terraform plan.
    - Assuming you configured an automatic trigger in your Jenkins Master (see [Jenkins sub-module README](../0-bootstrap/modules/jenkins-agent)), this will trigger a plan. You can also trigger a Jenkins job manually. Given the many options to do this in Jenkins, it is out of the scope of this document see [Jenkins website](http://www.jenkins.io) for more details.
   ```
   git push --set-upstream origin plan
   ```
1. Review the plan output in your Master's web UI.
1. Merge changes to production branch.
   ```
   git checkout -b production
   git push origin production
   ```
1. Review the apply output in your Master's web UI. (You might want to use the option to "Scan Multibranch Pipeline Now" in your Jenkins Master UI).

1. You can now move to the instructions in the step [2-environments](../2-environments/README.md).

### Deploying with Cloud Build

> 1. Clone the repo based on the Terraform output from the previous section. Run `terraform output cloudbuild_project_id` in the `0-bootstrap` folder to see the project again.

Reviewer 3: ```suggestion
1. Clone the policy repo based on the Terraform output from the previous section. Run `terraform output cloudbuild_project_id` in the `0-bootstrap` folder to see the project again.
```

   ```
   gcloud source repos clone gcp-policies --project=YOUR_CLOUD_BUILD_PROJECT_ID
   ```
1. Navigate into the repo.
   ```
   cd gcp-policies
   ```
1. Copy contents of policy-library to new repo.
   ```
   cp -RT ../terraform-example-foundation/policy-library/ .
   ```

1. Commit changes.
   ```

>    git add .` and `git commit -m 'Your message'

Reviewer 3: ```suggestion
   git add .
   git commit -m 'Your message'
```

>    ``

Reviewer 3: ```suggestion
   ```
```

   ```
1. Copy the Terraform wrapper script to the root of your new repository (modify accordingly based on your current directory).
   ```
   cp ../terraform-example-foundation/build/tf-wrapper.sh .
   ```
1. Ensure wrapper script can be executed.
   ```
   chmod 755 ./tf-wrapper.sh
   ```
1. Check if your organization already has a Access Context Manager Policy.
   ```
   gcloud access-context-manager policies list --organization YOUR_ORGANIZATION_ID --format="value(name)"
   ```
1. Rename `./envs/shared/terraform.example.tfvars` to `./envs/shared/terraform.tfvars` and update the file with values from your environment and bootstrap step (you can re-run `terraform output` in the 0-bootstrap directory to find these values). Make sure that `default_region` is set to a valid [BigQuery dataset region](https://cloud.google.com/bigquery/docs/locations). Also if the previous step showed a numeric value, make sure to un-comment the variable `create_access_context_manager_access_policy = false`. See the shared folder [README.md](./envs/shared/README.md) for additional information on the values in the `terraform.tfvars` file.
1. Commit changes.
   ```

>    git add .` and `git commit -m 'Your message'

Reviewer 3: ```suggestion
   git add .
   git commit -m 'Your message'
```

OS Login has some [limitations](https://cloud.google.com/compute/docs/instances/managing-instance-access#limitations).

> If those limitations do not apply to your workload/environment you can choose to enable the OS Login policy by setting variable `enable_os_login_policy` to `true`.

Reviewer: ```suggestion
If those limitations do not apply to your workload/environment, you can choose to enable the OS Login policy by setting variable `enable_os_login_policy` to `true`.
```

### Deploying with Jenkins

1. Clone the repo you created manually in 0-bootstrap.
   ```
   git clone <YOUR_NEW_REPO-1-org>
   ```

> 1. Navigate into the repo and change to a non production branch.

Reviewer: ```suggestion
1. Navigate into the repo and change to a non-production branch.
```

   ```
   cd YOUR_NEW_REPO_CLONE-1-org
   git checkout -b plan
   ```
1. Copy contents of foundation to new repo.
   ```
   cp -RT ../terraform-example-foundation/1-org/ .
   ```
1. Copy contents of policy-library to new repo.
   ```
   cp -RT ../terraform-example-foundation/policy-library/ ./policy-library
   ```
1. Copy the Jenkinsfile script to the root of your new repository.\

   ```
   cp ../terraform-example-foundation/build/Jenkinsfile .
   ```
1. Update the variables located in the `environment {}` section of the `Jenkinsfile` with values from your environment:
    ```
    _TF_SA_EMAIL
    _STATE_BUCKET_NAME
    _PROJECT_ID (the cicd project id)
    ```
1. Copy Terraform wrapper script to the root of your new repository.
   ```
   cp ../terraform-example-foundation/build/tf-wrapper.sh .
   ```
1. Ensure wrapper script can be executed.
   ```
   chmod 755 ./tf-wrapper.sh
   ```

> 1. Check if your organization already has a Access Context Manager Policy.

Reviewer: ```suggestion
1. Check if your organization already has an Access Context Manager Policy.
```

   ```
   gcloud access-context-manager policies list --organization YOUR_ORGANIZATION_ID --format="value(name)"
   ```

> 1. Rename `./envs/shared/terraform.example.tfvars` to `./envs/shared/terraform.tfvars` and update the file with values from your environment and bootstrap. You can re-run `terraform output` in the 0-bootstrap directory to find these values. Make sure that `default_region` is set to a valid [BigQuery dataset region](https://cloud.google.com/bigquery/docs/locations). Also if the previous step showed a numeric value, make sure to un-comment the variable `create_access_context_manager_access_policy = false`. See the shared folder [README.md](./envs/shared/README.md) for additional information on the values in the `terraform.tfvars` file.

Reviewer: ```suggestion
1. Rename `./envs/shared/terraform.example.tfvars` to `./envs/shared/terraform.tfvars` and update the file with values from your environment and bootstrap. You can re-run `terraform output` in the 0-bootstrap directory to find these values. Make sure that `default_region` is set to a valid [BigQuery dataset region](https://cloud.google.com/bigquery/docs/locations). Also, if the previous step showed a numeric value, make sure to un-comment the variable `create_access_context_manager_access_policy = false`. See the shared folder [README.md](./envs/shared/README.md) for additional information on the values in the `terraform.tfvars` file.
```

1. Commit changes.
   ```
   git add .
   git commit -m 'Your message'
   ```

> 1. Push your plan branch. The branch `plan` is not a special one. Any branch which name is different from `development`, `non-production` or `production` will trigger a terraform plan.

Reviewer: ```suggestion
1. Push your plan branch. The branch `plan` is not a special one. Any branch which name is different from `development`, `non-production` or `production` will trigger a Terraform plan.
```

    - Assuming you configured an automatic trigger in your Jenkins Master (see [Jenkins sub-module README](../0-bootstrap/modules/jenkins-agent)), this will trigger a plan. You can also trigger a Jenkins job manually. Given the many options to do this in Jenkins, it is out of the scope of this document see [Jenkins website](http://www.jenkins.io) for more details.
   ```
   git push --set-upstream origin plan
   ```
1. Review the plan output in your Master's web UI.
1. Merge changes to production branch.
   ```
   git checkout -b production
   git push origin production
   ```

> 1. Review the apply output in your Master's web UI. (You might want to use the option to "Scan Multibranch Pipeline Now" in your Jenkins Master UI).

Reviewer: ```suggestion
1. Review the apply output in your Master's web UI. (you might want to use the option to "Scan Multibranch Pipeline Now" in your Jenkins Master UI).
```

> 1. You can now move to the instructions in the step [2-environments](../2-environments/README.md).

Reviewer: ```suggestion
1. You can now move to the instructions in the [2-environments](../2-environments/README.md) step.
```

### Deploying with Cloud Build

1. Clone the policy repo based on the Terraform output from the previous section. Run `terraform output cloudbuild_project_id` in the `0-bootstrap` folder to see the project again.
   ```
   gcloud source repos clone gcp-policies --project=YOUR_CLOUD_BUILD_PROJECT_ID
   ```
1. Navigate into the repo.
   ```
   cd gcp-policies
   ```
1. Copy contents of policy-library to new repo.
   ```
   cp -RT ../terraform-example-foundation/policy-library/ .
   ```

1. Commit changes.
   ```
   git add .
   git commit -m 'Your message'
   ```
1. Push your master branch to the new repo.
   ```
   git push --set-upstream origin master
   ```
1. Navigate out of the repo.
   ```
   cd ..
   ```
1. Clone the repo.
   ```
   gcloud source repos clone gcp-org --project=YOUR_CLOUD_BUILD_PROJECT_ID
   ```

> 1. Navigate into the repo and change to a non production branch.

Reviewer: ```suggestion
1. Navigate into the repo and change to a non-production branch.
```

   ```
   cd gcp-org
   git checkout -b plan
   ```
1. Copy contents of foundation to new repo.
   ```
   cp -RT ../terraform-example-foundation/1-org/ .
   ```

> 1. Copy cloud build configuration files for Terraform. You may need to modify the command to reflect

Reviewer: ```suggestion
1. Copy Cloud Build configuration files for Terraform. You may need to modify the command to reflect
```

   your current directory.
   ```
   cp ../terraform-example-foundation/build/cloudbuild-tf-* .
   ```
1. Copy the Terraform wrapper script to the root of your new repository (modify accordingly based on your current directory).
   ```
   cp ../terraform-example-foundation/build/tf-wrapper.sh .
   ```
1. Ensure wrapper script can be executed.
   ```
   chmod 755 ./tf-wrapper.sh
   ```

> 1. Check if your organization already has a Access Context Manager Policy.

Reviewer: ```suggestion
1. Check if your organization already has an Access Context Manager Policy.
```

   ```
   gcloud access-context-manager policies list --organization YOUR_ORGANIZATION_ID --format="value(name)"
   ```

> 1. Rename `./envs/shared/terraform.example.tfvars` to `./envs/shared/terraform.tfvars` and update the file with values from your environment and bootstrap step (you can re-run `terraform output` in the 0-bootstrap directory to find these values). Make sure that `default_region` is set to a valid [BigQuery dataset region](https://cloud.google.com/bigquery/docs/locations). Also if the previous step showed a numeric value, make sure to un-comment the variable `create_access_context_manager_access_policy = false`. See the shared folder [README.md](./envs/shared/README.md) for additional information on the values in the `terraform.tfvars` file.

Reviewer: ```suggestion
1. Rename `./envs/shared/terraform.example.tfvars` to `./envs/shared/terraform.tfvars` and update the file with values from your environment and bootstrap step (you can re-run `terraform output` in the 0-bootstrap directory to find these values). Make sure that `default_region` is set to a valid [BigQuery dataset region](https://cloud.google.com/bigquery/docs/locations). Also, if the previous step showed a numeric value, make sure to un-comment the variable `create_access_context_manager_access_policy = false`. See the shared folder [README.md](./envs/shared/README.md) for additional information on the values in the `terraform.tfvars` file.
```

1. Commit changes.
   ```
   git add .
   git commit -m 'Your message'
   ```

> 1. Push your plan branch to trigger a plan. For this command, the branch `plan` is not a special one. Any branch which name is different from `development`, `non-production` or `production` will trigger a terraform plan.

Reviewer: ```suggestion
1. Push your plan branch to trigger a plan. For this command, the branch `plan` is not a special one. Any branch which name is different from `development`, `non-production` or `production` will trigger a Terraform plan.
```

   ```
   git push --set-upstream origin plan
   ```

> 1. Review the plan output in your cloud build project. https://console.cloud.google.com/cloud-build/builds?project=YOUR_CLOUD_BUILD_PROJECT_ID

Reviewer: ```suggestion
1. Review the plan output in your Cloud Build project. https://console.cloud.google.com/cloud-build/builds?project=YOUR_CLOUD_BUILD_PROJECT_ID
```

1. Merge changes to production branch.
   ```
   git checkout -b production
   git push origin production
   ```

> 1. Review the apply output in your cloud build project. https://console.cloud.google.com/cloud-build/builds?project=YOUR_CLOUD_BUILD_PROJECT_ID

Reviewer: ```suggestion
1. Review the apply output in your Cloud Build project. https://console.cloud.google.com/cloud-build/builds?project=YOUR_CLOUD_BUILD_PROJECT_ID
```

### Running Terraform locally

1. Change into 1-org folder.
1. Run `cp ../build/tf-wrapper.sh .`
1. Run `chmod 755 ./tf-wrapper.sh`
1. Change into 1-org/envs/shared/ folder.

> 1. Rename terraform.example.tfvars to terraform.tfvars and update the file with values from your environment and bootstrap.

Reviewer: ```suggestion
1. Rename `terraform.example.tfvars` to `terraform.tfvars` and update the file with values from your environment and bootstrap.
```

1. Obtain your bucket name by running the following command in the 0-bootstap folder.
   ```
   terraform output gcs_bucket_tfstate
   ```

> 1. Update backend.tf with your bucket from bootstrap.

Reviewer: ```suggestion
1. Update `backend.tf` with your bucket from bootstrap.
```

   ```
   for i in `find -name 'backend.tf'`; do sed -i 's/UPDATE_ME/<YOUR-BUCKET-NAME>/' $i; done
   ```

We will now deploy our environment (production) using this script.
When using Cloud Build or Jenkins as your CI/CD tool each environment corresponds to a branch is the repository for 1-org step and only the corresponding environment is applied.

> To use the `validate` option of the `tf-wrapper.sh` script, the latest version of `terraform-validator` must be [installed](https://github.com/forseti-security/policy-library/blob/master/docs/user_guide.md#how-to-use-terraform-validator) in your system and in you `PATH`.

Reviewer 3: ```suggestion
To use the `validate` option of the `tf-wrapper.sh` script, the latest version of `terraform-validator` must be [installed](https://github.com/forseti-security/policy-library/blob/master/docs/user_guide.md#how-to-use-terraform-validator) in your system and in your `PATH`.
```
